File: src/iterators/union_iterator.c
/*
 * Copyright (c) 2006-Present, Redis Ltd.
 * All rights reserved.
 *
 * Licensed under your choice of the Redis Source Available License 2.0
 * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
 * GNU Affero General Public License v3 (AGPLv3).
*/

#include "union_iterator.h"
#include "wildcard_iterator.h"
#include "iterators_rs.h"

static inline int cmpLastDocId(const void *e1, const void *e2, const void *udata) {
  const QueryIterator *it1 = e1, *it2 = e2;
  if (it1->lastDocId < it2->lastDocId)
    return 1;
  if (it1->lastDocId > it2->lastDocId)
    return -1;
  return 0;
}

static inline void resetMinIdHeap(UnionIterator *ui) {
  heap_t *hp = ui->heap_min_id;
  heap_clear(hp);
  for (int i = 0; i < ui->num_orig; i++) {
    if (!ui->its_orig[i]->atEOF) {
      heap_offerx(hp, ui->its_orig[i]);
    }
  }
}

static inline void UI_AddChild(UnionIterator *ui, QueryIterator *it) {
  AggregateResult_AddChild(ui->base.current, it->current);
}

/**
 * Removes the exhausted iterator from the active list, so that future
 * reads will no longer iterate over it
 */
static inline void UI_RemoveExhausted(UnionIterator *it, int idx) {
  // Quickly remove the iterator by swapping it with the last iterator.
  RS_ASSERT(0 <= idx && idx < it->num);
  it->its[idx] = it->its[--it->num]; // Also decrement the number of iterators
}

void UI_SyncIterList(UnionIterator *ui) {
  ui->num = ui->num_orig;
  memcpy(ui->its, ui->its_orig, sizeof(*ui->its) * ui->num_orig);
  if (ui->heap_min_id) {
    resetMinIdHeap(ui);
  }
  for (size_t i = 0; i < ui->num; i++) {
    while (i < ui->num && ui->its[i]->atEOF) {
      UI_RemoveExhausted(ui, i);
    }
  }
  if (ui->num == 0) {
    // If no active iterators, set EOF
    ui->base.atEOF = true;
  }
}

static size_t UI_NumEstimated(const QueryIterator *base) {
  const UnionIterator *ui = (const UnionIterator *)base;
  size_t estimation = 0;
  for (size_t i = 0; i < ui->num_orig; ++i) {
    estimation += ui->its_orig[i]->NumEstimated(ui->its_orig[i]);
  }
  return estimation;
}

static void UI_Rewind(QueryIterator *base) {
  UnionIterator *ui = (UnionIterator *)base;
  base->atEOF = false;
  base->lastDocId = 0;

  // rewind all child iterators
  for (size_t i = 0; i < ui->num_orig; i++) {
    ui->its_orig[i]->Rewind(ui->its_orig[i]);
  }

  UI_SyncIterList(ui);
}

// Collect all children whose current result is `ui->base.lastDocId`.
// Assumes that ui->base.current is already reset
static inline void UI_SetFullFlat(UnionIterator *ui) {
  for (int i = 0; i < ui->num; i++) {
    QueryIterator *cur = ui->its[i];
    if (cur->lastDocId == ui->base.lastDocId) {
      UI_AddChild(ui, cur);
    }
  }
}

// Set `ui` state (lastId, current result) according to a single child iterator
static inline void UI_QuickSet(UnionIterator *ui, QueryIterator *match) {
  ui->base.lastDocId = match->lastDocId;
  UI_AddChild(ui, match);
}

// Skip implementation, for full (no quick exit) mode, using an array.
// In full mode, we should never have child iterators lagging behind the union's last result
static inline IteratorStatus UI_Skip_Full_Flat(QueryIterator *base, const t_docId nextId) {
  RS_ASSERT(base->lastDocId < nextId);
  UnionIterator *ui = (UnionIterator *)base;
  if (base->atEOF) {
    return ITERATOR_EOF;
  }
  t_docId minId = DOCID_MAX;
  IndexResult_ResetAggregate(ui->base.current);
  for (int i = 0; i < ui->num; i++) {
    QueryIterator *cur = ui->its[i];
    if (cur->lastDocId < nextId) {
      IteratorStatus rc = cur->SkipTo(cur, nextId);
      if (rc == ITERATOR_OK) {
        UI_AddChild(ui, cur);
      } else if (rc == ITERATOR_NOTFOUND) {
        // No match. Finish the loop iteration normally
      } else if (rc == ITERATOR_EOF) {
        UI_RemoveExhausted(ui, i);
        i--;
        continue;
      } else {
        return rc;
      }
    } else if (cur->lastDocId == nextId) {
      UI_AddChild(ui, cur);
    }
    // Look for the minimal lastDocId
    if (minId > cur->lastDocId) minId = cur->lastDocId;
  }

  if (ui->num == 0) {
    base->atEOF = true;
    return ITERATOR_EOF;
  } else if (minId == nextId) {
    // We found what we were looking for
    ui->base.lastDocId = minId;
    // Current record was already set while scanning the children
    return ITERATOR_OK;
  } else {
    // We didn't find the requested ID, but we know the next minimal ID
    base->lastDocId = minId;
    UI_SetFullFlat(ui);
    return ITERATOR_NOTFOUND;
  }
}

// Read implementation, for full (no quick exit) mode, using an array.
// In full mode, we should never have child iterators lagging behind the union's last result
static inline IteratorStatus UI_Read_Full_Flat(QueryIterator *base) {
  UnionIterator *ui = (UnionIterator *)base;
  if (base->atEOF) {
    return ITERATOR_EOF;
  }
  const t_docId lastId = ui->base.lastDocId;
  t_docId minId = DOCID_MAX;
  IndexResult_ResetAggregate(ui->base.current);
  for (int i = 0; i < ui->num; i++) {
    QueryIterator *cur = ui->its[i];
    RS_LOG_ASSERT(cur->lastDocId >= lastId,
      "When reading in full mode, we should never have children behind the union last result");
    if (cur->lastDocId == lastId) {
      IteratorStatus rc = cur->Read(cur);
      if (rc == ITERATOR_OK) {
        // Finish the loop iteration normally
      } else if (rc == ITERATOR_EOF) {
        UI_RemoveExhausted(ui, i);
        i--;
        continue;
      } else {
        return rc;
      }
    }
    if (minId > cur->lastDocId) minId = cur->lastDocId;
  }
  if (ui->num) {
    base->lastDocId = minId;
    UI_SetFullFlat(ui);
    return ITERATOR_OK;
  } else {
    base->atEOF = true;
    return ITERATOR_EOF;
  }
}

// Skip implementation, for quick exit mode, using an array.
// In quick mode, we may have child iterators lagging behind the union's last result, so we cannot assume otherwise
static inline IteratorStatus UI_Skip_Quick_Flat(QueryIterator *base, const t_docId nextId) {
  RS_ASSERT(base->lastDocId < nextId);
  UnionIterator *ui = (UnionIterator *)base;
  if (base->atEOF) {
    return ITERATOR_EOF;
  }
  t_docId minId = DOCID_MAX;
  QueryIterator *minIt;
  IndexResult_ResetAggregate(ui->base.current);
  // TODO: performance improvement?
  // We can avoid performing any `SkipTo` call if we first scan an check if we already have a child
  // that matches the required ID.
  for (int i = 0; i < ui->num; i++) {
    QueryIterator *cur = ui->its[i];
    IteratorStatus rc;
    if (cur->lastDocId < nextId) {
      // Actually read from the child
      rc = cur->SkipTo(cur, nextId);
    } else if (cur->lastDocId == nextId) {
      // Current iterator previously matched this ID
      rc = ITERATOR_OK;
    } else { // (cur->lastDocId > nextId)
      // Current iterator previously advanced passed this ID
      rc = ITERATOR_NOTFOUND;
    }
    switch (rc) {
      case ITERATOR_OK:
        UI_QuickSet(ui, cur);
        return ITERATOR_OK;
      case ITERATOR_NOTFOUND:
        // Look for the minimal lastDocId + its iterator
        if (minId > cur->lastDocId) {
          minId = cur->lastDocId;
          minIt = cur;
        }
        break;
      case ITERATOR_EOF:
        UI_RemoveExhausted(ui, i);
        i--;
        break;
      case ITERATOR_TIMEOUT:
        return ITERATOR_TIMEOUT;
    }
  }

  if (ui->num) {
    // We didn't find the requested ID, but we set `minIt` to the next best match.
    UI_QuickSet(ui, minIt);
    return ITERATOR_NOTFOUND;
  } else {
    base->atEOF = true;
    return ITERATOR_EOF;
  }
}

// Read implementation, for quick exit mode, using an array.
// In quick mode, we may have child iterators lagging behind the union's last result, so we cannot assume otherwise.
// Therefore we need to use the children `SkipTo` to quickly get them to where we are. This is exactly
// what we already so in the `SkipTo` implemented above, so we can attempt to skip to the next possible ID.
// We just need to make sure to return `OK` even if we got `NOTFOUND` from the skipping step.
static inline IteratorStatus UI_Read_Quick_Flat(QueryIterator *base) {
  IteratorStatus rc = UI_Skip_Quick_Flat(base, base->lastDocId + 1);
  return rc == ITERATOR_NOTFOUND ? ITERATOR_OK : rc;
}

// Skip implementation, for full (no quick exit) mode, using a heap (many children).
// In full mode, we should never have child iterators lagging behind the union's last result
static inline IteratorStatus UI_Skip_Full_Heap(QueryIterator *base, const t_docId nextId) {
  RS_ASSERT(base->lastDocId < nextId);
  UnionIterator *ui = (UnionIterator *)base;
  if (base->atEOF) {
    return ITERATOR_EOF;
  }
  QueryIterator *cur;
  heap_t *hp = ui->heap_min_id;
  IndexResult_ResetAggregate(ui->base.current);
  while ((cur = heap_peek(hp)) && cur->lastDocId < nextId) {
    IteratorStatus rc = cur->SkipTo(cur, nextId);
    if (rc == ITERATOR_OK || rc == ITERATOR_NOTFOUND) {
      heap_replace(hp, cur); // replace current iterator with itself to update its position
    } else if (rc == ITERATOR_EOF) {
      heap_poll(hp);
    } else {
      return rc;
    }
  }

  if (cur) {
    // Set the current union result according to `cur`
    ui->base.lastDocId = cur->lastDocId;
    heap_cb_root(hp, (HeapCallback)UI_AddChild, ui); // Collect all matching children
    return nextId == cur->lastDocId ? ITERATOR_OK : ITERATOR_NOTFOUND;
  }
  base->atEOF = true;
  return ITERATOR_EOF;
}

// Read implementation, for full (no quick exit) mode, using a heap (many children).
// In full mode, we should never have child iterators lagging behind the union's last result,
// So we only need a single read on each child that matched on the previous read/skip call.
static inline IteratorStatus UI_Read_Full_Heap(QueryIterator *base) {
  UnionIterator *ui = (UnionIterator *)base;
  if (base->atEOF) {
    return ITERATOR_EOF;
  }
  QueryIterator *cur;
  heap_t *hp = ui->heap_min_id;
  IndexResult_ResetAggregate(ui->base.current);
  while ((cur = heap_peek(hp)) && cur->lastDocId == base->lastDocId) {
    IteratorStatus rc = cur->Read(cur);
    if (rc == ITERATOR_OK) {
      heap_replace(hp, cur); // replace current iterator with itself to update its position
    } else if (rc == ITERATOR_EOF) {
      heap_poll(hp);
    } else {
      return rc;
    }
  }

  if (cur) {
    // Set the current union result according to `cur`
    RS_ASSERT(cur->lastDocId > base->lastDocId);
    ui->base.lastDocId = cur->lastDocId;
    heap_cb_root(hp, (HeapCallback)UI_AddChild, ui); // Collect all matching children
    return ITERATOR_OK;
  }
  base->atEOF = true;
  return ITERATOR_EOF;
}

// Skip implementation, for quick exit mode, using a heap (many children).
// In quick mode, we may have child iterators lagging behind the union's last result.
// This implementation is very similar to `UI_Skip_Full_Heap`, beside the quick exit when
// finding an exact match.
static inline IteratorStatus UI_Skip_Quick_Heap(QueryIterator *base, const t_docId nextId) {
  RS_ASSERT(base->lastDocId < nextId);
  UnionIterator *ui = (UnionIterator *)base;
  if (base->atEOF) {
    return ITERATOR_EOF;
  }
  QueryIterator *cur;
  heap_t *hp = ui->heap_min_id;
  IndexResult_ResetAggregate(ui->base.current);
  // TODO: performance improvement?
  // before attempting to advance lagging iterators, we can perform a quick scan of the heap and check if
  // we already have a matching iterator, saving us from performing any `SkipTo` call, which may be expensive.
  while ((cur = heap_peek(hp)) && cur->lastDocId < nextId) {
    IteratorStatus rc = cur->SkipTo(cur, nextId);
    if (rc == ITERATOR_OK) {
      heap_replace(hp, cur); // replace current iterator with itself to update its position
      UI_QuickSet(ui, cur);
      return ITERATOR_OK;
    } else if (rc == ITERATOR_NOTFOUND) {
      heap_replace(hp, cur); // replace current iterator with itself to update its position
    } else if (rc == ITERATOR_EOF) {
      heap_poll(hp);
    } else {
      return rc;
    }
  }

  if (cur) {
    UI_QuickSet(ui, cur);
    return nextId == cur->lastDocId ? ITERATOR_OK : ITERATOR_NOTFOUND;
  }
  base->atEOF = true;
  return ITERATOR_EOF;
}

// Read implementation, for quick exit mode, using a heap.
// In quick mode, we may have child iterators lagging behind the union's last result, so we cannot assume otherwise.
// Therefore we need to use the children `SkipTo` to quickly get them to where we are. This is exactly
// what we already so in the `SkipTo` implemented above, so we can attempt to skip to the next possible ID.
// We just need to make sure to return `OK` even if we got `NOTFOUND` from the skipping step.
static inline IteratorStatus UI_Read_Quick_Heap(QueryIterator *base) {
  IteratorStatus rc = UI_Skip_Quick_Heap(base, base->lastDocId + 1);
  return rc == ITERATOR_NOTFOUND ? ITERATOR_OK : rc;
}

static void UI_Free(QueryIterator *base) {
  if (base == NULL) return;

  UnionIterator *ui = (UnionIterator *)base;
  for (int i = 0; i < ui->num_orig; i++) {
    QueryIterator *it = ui->its_orig[i];
    if (it) {
      it->Free(it);
    }
  }

  IndexResult_Free(ui->base.current);
  if (ui->heap_min_id) heap_free(ui->heap_min_id);
  rm_free(ui->its);
  rm_free(ui->its_orig);
  rm_free(ui);
}

/**
 * Reduce the union iterator by applying these rules:
 * 1. Remove all empty iterators
 * 2. If in quick exit mode and any of the iterators is a wildcard iterator, return it and free the rest
 * 3. Otherwise, return NULL and let the caller create the union iterator
 */
static QueryIterator *UnionIteratorReducer(QueryIterator **its, int *num, bool quickExit) {
  QueryIterator *ret = NULL;
  // Let's remove all the empty iterators from the list
  size_t current_size = *num;
  size_t write_idx = 0;
  for (size_t i = 0; i < current_size; ++i) {
    if (its[i]) {
      if (its[i]->type != EMPTY_ITERATOR) {
        its[write_idx++] = its[i];
      } else {
        its[i]->Free(its[i]);
      }
    }
  }
  *num = write_idx;
  if (quickExit) {
    for (size_t i = 0; i < write_idx; ++i) {
      if (IsWildcardIterator(its[i])) {
        ret = its[i];
        for (size_t j = 0; j < write_idx; ++j) {
          if (i != j && its[j]) {
            its[j]->Free(its[j]);
          }
        }
        break;
      }
    }
  }

  if (write_idx == 1) {
    ret = its[0];
  } else if (write_idx == 0) {
    ret = NewEmptyIterator();
  }
  if (ret != NULL) {
    rm_free(its);
  }
  return ret;
}

static ValidateStatus UI_Revalidate(QueryIterator *base) {
  UnionIterator *ui = (UnionIterator *)base;
  t_docId original_lastDocId = base->lastDocId;
  bool all_child_ok = true;

  // Loop over the original iterators list and revalidate each child
  // Use read/write indexes to efficiently pack the array
  uint32_t new_num_orig = 0;
  for (uint32_t i = 0; i < ui->num_orig; i++) {
    QueryIterator *child = ui->its_orig[i];
    RS_ASSERT(child != NULL);

    ValidateStatus child_status = child->Revalidate(child);
    all_child_ok = all_child_ok && (child_status == VALIDATE_OK);

    if (child_status == VALIDATE_ABORTED) {
      // Free the aborted child
      child->Free(child);
      // Don't copy this child to write position (effectively removes it)
    } else {
      // Keep this child - copy it to write position
      ui->its_orig[new_num_orig++] = child;
    }
  }
  if (all_child_ok) {
    // No children were removed or moved, just return
    return VALIDATE_OK;
  }
  ui->num_orig = new_num_orig;
  if (ui->num_orig == 0) {
    // All children were aborted, we should abort the union iterator
    return VALIDATE_ABORTED;
  }

  // Use UI_SyncIterList to update `its` array and heap - remove exhausted iterators
  UI_SyncIterList(ui);

  // Update current result - reset and rebuild if we have active children
  IndexResult_ResetAggregate(ui->base.current);
  // Find the minimum docId among active children to update current result
  t_docId minId = DOCID_MAX;
  for (int i = 0; i < ui->num; i++) {
    if (ui->its[i]->lastDocId < minId) {
      minId = ui->its[i]->lastDocId;
    }
  }
  if (ui->num > 0) base->lastDocId = minId;
  // Set the current result based on the minimum docId found, regardless of quick exit or algorithm
  UI_SetFullFlat(ui);

  // Return VALIDATE_MOVED only if our lastDocId actually changed
  return (base->lastDocId != original_lastDocId) ? VALIDATE_MOVED : VALIDATE_OK;
}

QueryIterator *NewUnionIterator(QueryIterator **its, int num, bool quickExit,
                                double weight, QueryNodeType type, const char *q_str, IteratorsConfig *config) {

  QueryIterator* ret = UnionIteratorReducer(its, &num, quickExit);
  if (ret != NULL) {
    return ret;
  }
  // create union context
  UnionIterator *ui = rm_calloc(1, sizeof(UnionIterator));
  ui->its_orig = its;
  ui->type = type;
  ui->num_orig = num;
  ui->its = rm_malloc(num * sizeof(*ui->its));
  ui->heap_min_id = NULL;
  ui->q_str = q_str;

  // bind the union iterator calls
  ret = &ui->base;
  ret->type = UNION_ITERATOR;
  ret->atEOF = false;
  ret->lastDocId = 0;
  ret->current = NewUnionResult(num, weight);
  ret->NumEstimated = UI_NumEstimated;
  ret->Free = UI_Free;
  ret->Rewind = UI_Rewind;
  ret->Revalidate = UI_Revalidate;

  // Choose `Read` and `SkipTo` implementations.
  // We have 2 factors for the choice:
  // 1. quickExit - whether to return after the first match was found without checking if more iterator agree with it,
  //                or should we collect all the results from all the children that agree on the current ID.
  // 2. minUnionIterHeap - choose whether to use a flat array or a heap for tracking the children, according to the number of children
  // Each implementation if fine-tuned for the best performance in its scenario, and relies on the current state
  // of the iterator and how it was left by previous API calls, so we can't change implementation mid-execution.
  if (num > config->minUnionIterHeap) {
    ret->Read = quickExit ? UI_Read_Quick_Heap : UI_Read_Full_Heap;
    ret->SkipTo = quickExit ? UI_Skip_Quick_Heap : UI_Skip_Full_Heap;
    ui->heap_min_id = rm_malloc(heap_sizeof(num));
    heap_init(ui->heap_min_id, cmpLastDocId, NULL, num);
  } else {
    ret->Read = quickExit ? UI_Read_Quick_Flat : UI_Read_Full_Flat;
    ret->SkipTo = quickExit ? UI_Skip_Quick_Flat : UI_Skip_Full_Flat;
  }

  UI_SyncIterList(ui);
  return ret;
}

File: src/redisearch_rs/rqe_iterators/src/union_iterator.rs
/*
 * Copyright (c) 2006-Present, Redis Ltd.
 * All rights reserved.
 *
 * Licensed under your choice of the Redis Source Available License 2.0
 * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
 * GNU Affero General Public License v3 (AGPLv3).
*/

use crate::rqe_iterator::{RQEIterator, RQEIteratorError, RQEValidateStatus, SkipToOutcome};
use std::usize;
use ffi::t_docId;
use inverted_index::RSIndexResult;
use low_memory_thin_vec::LowMemoryThinVec;

enum UnionIteratorMode {
    /// Represents a union iterator that aggregates results from multiple child iterators.
    FlatUnion,
    /// Represents a union iterator that aggregates results from multiple child iterators, but does not flatten the results.
    HeapUnion {
        heap: std::collections::BinaryHeap<Box<dyn RQEIterator>>,
    },
}

/// Represents a union iterator that aggregates results from multiple child iterators
struct UnionIterator {
    /// The active (not depleted) children iterators that are currently being processed
    active_children: LowMemoryThinVec<Box<dyn RQEIterator>>,
    /// The children iterators that are not currently being processed, but can be used later
    /// This is used to avoid re-allocating memory for children iterators that are not currently active.
    children: LowMemoryThinVec<Box<dyn RQEIterator>>,
    /// The last document ID that was read from the iterator
    last_doc_id: t_docId,
    /// The current result of the iterator, which is the union of the results from the active children iterators
    current_result: RSIndexResult,
    /// Indicates if the instance of the iterator is a flat union iterator or a heap union iterator, and use enum invocations on specific methods.
    mode: UnionIteratorMode,
    /// Indicates if this iterator uses the quick implementations.
    quick: bool,
    /// Indicates if the iterator has more results to read.
    has_next: bool,
}

impl UnionIterator {
    // TODO: create a function to return Box<dyn RQEIterator> for union iterators, according to given parameters (Union iterator factory)
    // This should be concise with UnionIteratorReducer from the C code.
    pub fn factory(
        children: LowMemoryThinVec<Box<dyn RQEIterator>>,
        quick: bool,
    ) -> Box<dyn RQEIterator> {
        // Create a new union iterator with the given children iterators
        let mode = UnionIteratorMode::FlatUnion; // Default to FlatUnion, can be changed later
        Box::new(UnionIterator::new(children, mode, quick))
    }

    pub fn new(
        children: LowMemoryThinVec<Box<dyn RQEIterator>>,
        mode: UnionIteratorMode,
        quick: bool,
    ) -> Self {
        let active_children = LowMemoryThinVec::with_capacity(children.len());
        let current_result = RSIndexResult::union(0);
        let last_doc_id = 0;
        let has_next = !children.is_empty();

        UnionIterator {
            active_children,
            children,
            last_doc_id,
            current_result,
            mode,
            quick,
            has_next,
        }
    }

    /// Removes an exhausted child iterator from the active children list and puts it back to the children list.
    #[inline(always)]
    fn remove_exhausted_child(&mut self, idx: usize) {
        // Assert that the index is within bounds
        assert!(
            idx < self.active_children.len(),
            "Index out of bounds for union iterator children"
        );
        // Swap the exhausted child with the last active child and put it back to the children list.
        let child = self.active_children.swap_remove(idx);
        self.children.push(child);
    }

    fn remove_exhausted_children(&mut self) {
        // Remove all exhausted children iterators from the active children list and put them back to the children list.
        // This is used to avoid re-allocating memory for children iterators that are not currently active.
        // This is used when the iterator is being reset to the beginning.
        let mut idx = 0;
        while idx < self.active_children.len() {
            if !self.active_children[idx].has_next() {
                self.remove_exhausted_child(idx);
            } else {
                idx += 1;
            }
        }
    }

    /// Moves all children iterators to the active children list.
    #[inline(always)]
    fn move_to_active(&mut self) {
        // Move all children iterators to the active children list.
        // This is used to avoid re-allocating memory for children iterators that are not currently active.
        // This is used when the iterator is being initialized or when it is being reset
        // to the beginning.
        while !self.children.is_empty() {
            self.active_children.push(self.children.pop().unwrap());
        }
    }

    #[inline(always)]
    fn move_to_children(&mut self) {
        // Move all active children iterators to the children list.
        // This is used to avoid re-allocating memory for children iterators that are not currently active.
        // This is used when the iterator is being reset to the beginning.
        while !self.active_children.is_empty() {
            self.children.push(self.active_children.pop().unwrap());
        }
    }

    /// Synchronizes the active children list with the current state of the iterators.
    fn sync_iter_list(&mut self) {
        self.move_to_active();
        self.remove_exhausted_children();
        // If there are no active children, set has_next to false.
        self.has_next = !self.active_children.is_empty();
    }

    fn set_full_flat(&mut self, results: &LowMemoryThinVec<RSIndexResult>) {
        // Reset the current result
        // TODO: should we move `reset` to the RSIndexResult?
        let current_result = self.current_result.as_union_mut().unwrap();
        current_result.reset();
        // Iterate over all results and add them to the current result
        for result in results.iter() {
            if result.get_doc_id() == self.last_doc_id {
                // If the result is the minimum ID, we add it to the current result
                self.current_result.push(result);
            }
        }
    }

    fn read_full_flat(&mut self) -> Result<Option<RSIndexResult>, RQEIteratorError> {
        let last_id: t_docId = self.last_doc_id;
        // TODO: add doc id limits to rust
        let mut min_id: t_docId = u64::MAX as t_docId;

        let mut results = LowMemoryThinVec::with_capacity(self.active_children.len());
        // Iterate over all active children iterators
        for child in self.active_children.iter_mut() {
            assert!(
                child.last_doc_id() >= last_id,
                "Child iterator's last doc id is less than the union iterator's last doc id"
            );
            // Read the next result from the child iterator
            match child.read() {
                Ok(Some(result)) => {
                    min_id = min_id.min(result.get_doc_id());
                    results.push(result);
                }
                Ok(None) => {
                    // do nothing, the child iterator is exhausted and we will remove it later
                }
                Err(e) => {
                    return Err(e);
                }
            }
        }
        // Remove exhausted children iterators
        self.remove_exhausted_children();
        if results.is_empty() {
            // If there are no results, all active children are exhausted
            assert!(
                self.active_children.len() == 0,
                "Union iterator has no results, but still has active children"
            );
            // If there are no results, we return None, since we are at the end of the iterator
            self.has_next = false;
            return Ok(None);
        }

        self.last_doc_id = min_id;
        self.set_full_flat(&results);
        return Ok(Some(self.current_result.clone()));
    }

    fn skip_to_full_flat(
        &mut self,
        doc_id: t_docId,
    ) -> Result<Option<SkipToOutcome>, RQEIteratorError> {
        let mut min_id = u64::MAX as t_docId;
        let mut results = LowMemoryThinVec::with_capacity(self.active_children.len());
        for child in self.active_children.iter_mut() {
            assert!(
                child.last_doc_id() >= doc_id,
                "Child iterator's last doc id is less than the requested doc id"
            );
            match child.skip_to(doc_id) {
                Ok(Some(SkipToOutcome::Found(result))) => {
                    min_id = min_id.min(result.get_doc_id());
                    results.push(result);
                }
                Ok(Some(SkipToOutcome::NotFound(result))) => {
                    min_id = min_id.min(result.get_doc_id());
                    results.push(result);
                }
                Ok(None) => {
                    // do nothing, the child iterator is exhausted and we will remove it later
                }
                Err(e) => {
                    return Err(e);
                }
            }
        }
        // Remove exhausted children iterators
        self.remove_exhausted_children();
        if results.is_empty() {
            // if ther are no results, all active children are exhausted
            assert!(
                self.active_children.len() == 0,
                "Union iterator has no results, but still has active children"
            );
            // If there are no results, we return None, since we are at the end of the iterator
            self.has_next = false;
            return Ok(None);
        }

        // If there are results, we need to find the minimum ID and set the current result
        results.retain(|r| r.get_doc_id() == min_id);
        self.set_full_flat(&results);

        if min_id == doc_id {
            // If the minimum ID is equal to the requested doc id, we return the current result
            self.last_doc_id = min_id;
            Ok(Some(SkipToOutcome::Found(self.current_result.clone())))
        } else {
            // If the minimum ID is greater than the requested doc id, we return a NotFound outcome
            self.last_doc_id = min_id;
            Ok(Some(SkipToOutcome::NotFound(self.current_result.clone())))
        }
    }
}

impl RQEIterator for UnionIterator {
    fn read(&mut self) -> Result<Option<RSIndexResult>, RQEIteratorError> {
        if !self.has_next {
            return Ok(None);
        }
        match self.mode {
            UnionIteratorMode::FlatUnion => self.read_full_flat(),
            _ => Ok(None), // Placeholder for other modes, such as HeapUnion
        }
    }

    fn skip_to(&mut self, doc_id: t_docId) -> Result<Option<SkipToOutcome>, RQEIteratorError> {
        assert!(
            self.last_doc_id < doc_id,
            "Union iterator's last doc id is not less than the requested doc id"
        );
        // If the iterator has no next, we return None
        if !self.has_next {
            return Ok(None);
        }
        match self.mode {
            UnionIteratorMode::FlatUnion => self.skip_to_full_flat(doc_id),
            _ => Ok(None), // Placeholder for other modes, such as HeapUnion
        }
    }

    fn revalidate(&mut self) -> RQEValidateStatus {
        // First we will remove all aborted children iterators
        self.move_to_children();
        self.children
            .retain_mut(|child| child.revalidate() != RQEValidateStatus::Aborted);
        if self.children.is_empty() {
            // If there are no children, we are done
            self.has_next = false;
            return RQEValidateStatus::Aborted;
        }

        // Now we will move all children iterators to the active children list and remove exhausted children iterators
        self.sync_iter_list();
        if self.active_children.is_empty() {
            // If there are no active children, we are done
            self.has_next = false;
            return RQEValidateStatus::Ok;
        } else {
            self.has_next = true;
            let original_id = self.last_doc_id;
            self.last_doc_id = self
                .active_children
                .iter()
                .map(|c| c.last_doc_id())
                .min()
                .unwrap_or(u64::MAX as t_docId);
            if self.last_doc_id == original_id {
                // If the last doc id is the same as before, we are still at the same position
                return RQEValidateStatus::Ok;
            } else {
                // If the last doc id has changed, we have moved to a new position
                return RQEValidateStatus::Moved;
            }
        }
        // Unlike the C implementation, we will not call set_full_flat here, since we are not reading any results yet.
        // Instead, we will just reset the last_doc_id and has_next properties.
    }

    fn rewind(&mut self) {
        self.last_doc_id = 0;
        self.has_next = false;
        for child in self.active_children.iter_mut() {
            child.rewind();
        }
        for child in self.children.iter_mut() {
            child.rewind();
        }
        self.sync_iter_list();
        // TODO: should we reset the current result?
    }

    fn num_estimated(&self) -> usize {
        let mut sum: usize = 0;
        sum = sum.saturating_add(
            self.active_children
                .iter()
                .map(|c| c.num_estimated())
                .sum::<usize>(),
        );
        sum = sum.saturating_add(
            self.children
                .iter()
                .map(|c| c.num_estimated())
                .sum::<usize>(),
        );
        sum
    }

    fn last_doc_id(&self) -> t_docId {
        return self.last_doc_id;
    }

    fn has_next(&self) -> bool {
        self.has_next
    }
}
